softlayer/resource_softlayer_network_vlan.go

```diff
@@ -0,0 +1,419 @@
+				return nil, "pending", nil
+			} else {
+				return nil, "", fmt.Errorf("Expected one network vlan: %s", err)
+			}
+		},
+		Timeout:    10 * time.Minute,
+		Delay:      5 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	pendingResult, err := stateConf.WaitForState()
+
+	if err != nil {
+		return datatypes.Network_Vlan{}, err
+	}
+
+	var result, ok = pendingResult.(datatypes.Network_Vlan)
+
+	if ok {
+		return result, nil
+	}
+
+	return datatypes.Network_Vlan{},
+		fmt.Errorf("Cannot find Network Vlan with order id '%d'", orderId)
+}
+
+func buildVlanProductOrderContainer(d *schema.ResourceData, sess *session.Session, packageType string) (
+	*datatypes.Container_Product_Order_Network_Vlan, error) {
+	var dc datatypes.Location_Datacenter
+	var err error
+	var rt datatypes.Hardware
+	router := d.Get("primary_router_hostname").(string)
+
+	vlanType := d.Get("type").(string)
+	datacenter := d.Get("datacenter").(string)
+
+	if len(datacenter) > 0 {
+		dc, err = location.GetDatacenterByName(sess, datacenter, "id")
+		if err != nil {
+			return &datatypes.Container_Product_Order_Network_Vlan{}, err
+		}
+	} else {
+		return &datatypes.Container_Product_Order_Network_Vlan{},
+			fmt.Errorf("datacenter name is empty.")
+	}
+
+	// 1. Get a package
+	pkg, err := product.GetPackageByType(sess, packageType)
+	if err != nil {
+		return &datatypes.Container_Product_Order_Network_Vlan{}, err
+	}
+
+	// 2. Get all prices for the package
+	productItems, err := product.GetPackageProducts(sess, *pkg.Id)
+	if err != nil {
+		return &datatypes.Container_Product_Order_Network_Vlan{}, err
+	}
+
+	// 3. Find vlan and subnet prices
+	vlanKeyname := vlanType + "_NETWORK_VLAN"
+	subnetKeyname := strconv.Itoa(d.Get("primary_subnet_size").(int)) + "_STATIC_PUBLIC_IP_ADDRESSES"
+
+	// 4. Select items with a matching keyname
+	vlanItems := []datatypes.Product_Item{}
+	subnetItems := []datatypes.Product_Item{}
+	for _, item := range productItems {
+		if *item.KeyName == vlanKeyname {
+			vlanItems = append(vlanItems, item)
+		}
+		if strings.Contains(*item.KeyName, subnetKeyname) {
+			subnetItems = append(subnetItems, item)
+		}
+	}
+
+	if len(vlanItems) == 0 {
+		return &datatypes.Container_Product_Order_Network_Vlan{},
+			fmt.Errorf("No product items matching %s could be found", vlanKeyname)
+	}
+
+	if len(subnetItems) == 0 {
+		return &datatypes.Container_Product_Order_Network_Vlan{},
+			fmt.Errorf("No product items matching %s could be found", subnetKeyname)
+	}
+
+	productOrderContainer := datatypes.Container_Product_Order_Network_Vlan{
+		Container_Product_Order: datatypes.Container_Product_Order{
+			PackageId: pkg.Id,
+			Location:  sl.String(strconv.Itoa(*dc.Id)),
+			Prices: []datatypes.Product_Item_Price{
+				{
+					Id: vlanItems[0].Prices[0].Id,
+				},
+				{
+					Id: subnetItems[0].Prices[0].Id,
+				},
+			},
+			Quantity: sl.Int(1),
+		},
+	}
+
+	if len(router) > 0 {
+		rt, err = hardware.GetRouterByName(sess, router, "id")
```

Comment:
Right. At first, I thought that Hardware or Hardware_Router services would provide those functions. However, GetRouterByName does not use any Hareware_* services. Only a return type of GetRouterByName is `Hardware`. Except Router, other resources such as SoftLayer_Network_Pod and SoftLayer_Network_Backbone are using Network_* datatypes. So their helper functions will be located in network package in the future. So network package can be good place for GetRouterByName.